src/include/internal/utility/alloc_impl.h

```diff
@@ -0,0 +1,194 @@
+/**
+ * @file
+ *
+ * @brief Helper for memory management.
+ *
+ * This file contains the implementations for the functions declared in
+ * <elektra/utility/alloc.h> and <internal/utility/alloc.h>
+ *
+ * @copyright BSD License (see LICENSE.md or https://www.libelektra.org)
+ */
+
+#ifndef ELEKTRA_UTILITY_ALLOC_IMPL_H
+#define ELEKTRA_UTILITY_ALLOC_IMPL_H
+
+#include <internal/utility/assert.h>
+
+#include <limits.h>
+#include <stddef.h>
+#include <stdlib.h>
+#include <string.h>
+
+#ifdef __cplusplus
+namespace ckdb
+{
+extern "C" {
+#endif
+
+/**Reallocate Storage in a save way.
+ *
+ *@code
+if (elektraRealloc ((void **) & buffer, new_length) < 0) {
+	// here comes the failure handler
+	// you can still use the old buffer
+#if DEBUG
+	fprintf (stderr, "Reallocation error\n");
+#endif
+	elektraFree (buffer);
+	buffer = 0;
+	// return with error
+}
+ *@endcode
+ *
+ * @param buffer is a pointer to a elektraMalloc
+ * @param size is the new size for the memory
+ * @retval -1 on failure
+ * @retval 0 on success
+ * @ingroup internal
+ */
+int elektraRealloc (void ** buffer, size_t size)
+{
+	ELEKTRA_ASSERT (size, "Size to allocate is zero (implementation defined behavior)");
+	void * ptr;
+	void * svr = *buffer;
+
+	ptr = realloc (*buffer, size);
+	ELEKTRA_ASSERT (ptr, "Memory (re)allocation failed with size %zu", size);
+	if (ptr == NULL)
+	{
+		// TODO: is this safe? couldn't the old buffer already be destroyed?
+		*buffer = svr; /* restore old buffer*/
+		return -1;
+	}
+	else
+	{
+		*buffer = ptr;
+		return 0;
+	}
```

Comment:
```suggestion
	ptr = realloc (*buffer, size);
        if (ptr == NULL)
        {
            return -1;
        }
        *buffer = ptr;
        return 0;
```
AFAIK the behavior of realloc is that it only frees the old memory if it successfully allocates the new memory. [1] So, there's no need to revert `*buffer` back to `svr`. Instead, you can just return -1 directly.

[1] https://linux.die.net/man/3/realloc

Reply:
I actually just added the `TODO`, because I found the code odd. The rest is copied from the old `internal.c`. But you're right, the `svr` stuff is completely unnecessary. Not just because of how `realloc` works, but simply because `realloc` has no way of changing the value of `*buffer`. It could only change the data in `**buffer`.

However, your suggestion would make `svr` an unused variable, so I won't use it directly.